### ISSUE TITLE
Bug in parsing POST message with JSON objects (server side)

### ISSUE THREAD

**Issue by yarongot** [pistacheio/pistache#94]
In http.cc function "BodyStep::apply" the cursor is being advanced by 2 to skip the CRLF, but it is not promised that the values being skipped are actually CRLF. In fact in my test, the 2 characters where actually the beginning of the JSON object that was skipped.

So I tweaked this to the following:

        if (message->body_.empty()) {
            /* If this is the first time we are reading the body, skip the CRLF */
        	switch(cursor.current())
        	{
        	case 10:
        	case 13:
        		if (!cursor.advance(2))
        			return State::Again;
        		break;
        	}

            //if (!cursor.advance(2)) return State::Again;
        }

**Comment by nelsonhf** [comment]
The definition of HTTP requires an empty line between the header and the body.
At this point, if "this is the first time we are reading the body" there MUST be a CR followed by an LF. If the pair is not there, the HTTP is not well-formed.

According to [IETF RFC 7230, section 3 (Message Format)](https://tools.ietf.org/html/rfc7230#section-3):

> All HTTP/1.1 messages consist of a start-line followed by a sequence
>    of octets in a format similar to the Internet Message Format
>    [RFC5322]: zero or more header fields (collectively referred to as
>    the "headers" or the "header section"), an empty line indicating the
>    end of the header section, and an optional message body.
> 
>      HTTP-message   = start-line
>                       *( header-field CRLF )
>                       CRLF
>                       [ message-body ]

**Comment by nelsonhf** [comment]
I found what could be a better solution for this problem. Instead of considering the "empty line" as the first part of the body, consider it as the last part of the header.
In terms of code:
Add the following line to right before the `return State::Next;` in function `HeadersStep::apply(StreamCursor& cursor)` (line 317? in http.cc):
```
        if (!cursor.advance(2)) return State::Again;
```
Remove the following block from the top of function `BodyStep::apply(StreamCursor& cursor)` (line 322? in http.cc):
```
        if (message->body_.empty()) {
            /* If this is the first time we are reading the body, skip the CRLF */
            if (!cursor.advance(2)) return State::Again;
        }

```
I believe (but cannot prove) that the problem is caused by the "empty line" being the last bytes read. Then they will be skipped but nothing would have been added to the body, which will be empty in the next call, causing the first 2 characters in the actual body to be skipped.
This is the behavior I was seeing and probably what yarongot (and others) experienced, too.
